### PR TITLE
Repair tow problems of endpoint and timeout

### PR DESCRIPTION
**Pull request by xiaomoli**
1. Modify the way of golangsdk.ServiceClient  --endpoint
2. Set up external_provisioner `--timeout=80s`  --timeout
3. Add idempotency for creating volumes   --timeout
4. etc

### CODE REVIEW

File: pkg/sfs/controllerserver.go [hunk 1]
```diff
@@ -39,47 +44,62 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
+	// check pv hasCreate or not
+	if value, ok := pvcProcessSuccess[req.Name]; ok && value != nil {
```
Comment (chengxiangdong) [inline pkg/sfs/controllerserver.go:48]: Checking for existence with `pvcProcessSuccess` may caust bugs,
it is better to call the service to check for existence.

Comment (xiaomoli) [reply pkg/sfs/controllerserver.go]: Modifications have been completed

File: go.mod [hunk 2]
```diff
@@ -1,23 +1,27 @@
...
 	github.com/spf13/pflag v1.0.5
 	github.com/unknwon/com v1.0.1
 	golang.org/x/net v0.0.0-20191004110552-13f9640d40b9
+	google.golang.org/grpc v1.21.0
+	gopkg.in/gcfg.v1 v1.2.3
+	k8s.io/klog v1.0.0
+)
+
+require (
+	github.com/golang/protobuf v1.3.2 // indirect
```
Comment (chengxiangdong) [inline go.mod:21]: there are too many unnecessary changes, please do some deletions

File: pkg/sfs/controllerserver.go [hunk 3]
```diff
@@ -40,46 +45,64 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
 	client, err := cs.Driver.cloud.SFSV2Client()
-    if err != nil {
+	if err != nil {
 		klog.V(3).Infof("Failed to create SFS v2 client: %v", err)
 		return nil, status.Error(codes.Internal, err.Error())
-    }
+	}
+
+	// check pv hasCreate or not
+	if value, ok := pvcProcessSuccess[req.Name]; ok && value != nil {
```
Comment (chengxiangdong) [inline pkg/sfs/controllerserver.go:54]: why use `pvcProcessSuccess` ? The data in memory is unreliable, it will lose after restarting

File: pkg/sfs/controllerserver.go [hunk 4]
```diff
@@ -91,16 +114,20 @@ func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
...
-    }
+	}
 	err = deleteShare(client, volID)
 	if err != nil {
 		klog.V(3).Infof("Failed to DeleteVolume: %v", err)
 		return nil, status.Error(codes.Internal, fmt.Sprintf("DeleteVolume failed with error %v", err))
 	}
 
+	////delete cache
+	pvcID := shareIDGetPvcID[volID]
```
Comment (chengxiangdong) [inline pkg/sfs/controllerserver.go:128]: do not use `shareIDGetPvcID` and `pvcProcessSuccess` it is unreliable

File: pkg/sfs/share.go [hunk 5]
```diff
@@ -25,23 +25,14 @@ import (
...
-	if err != nil {
-		return nil, err
-	}
-
-	err = waitForShareStatus(client, share.ID, shareAvailable, waitForAvailableShareTimeout)
-	if err != nil {
-		return nil, err
-	}
-	return share, nil
+	createOpts.Description = shareDescription
```
Comment (chengxiangdong) [inline pkg/sfs/share.go:34]: `createShare` contains logic to wait, please do not change its definition

File: pkg/sfs/controllerserver.go [hunk 6]
```diff
@@ -40,46 +45,64 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
 
-    share, err := createShare(client, &createOpts)
+	share, err := createShare(client, &createOpts)
 	if err != nil {
 		klog.V(3).Infof("Failed to create SFS volume: %v", err)
 		return nil, status.Error(codes.Internal, err.Error())
-    }
+	}
+
+	volume := &csi.Volume{
```
[the remark this answers is not in the document]
Comment (chengxiangdong) [reply pkg/sfs/controllerserver.go:83]: the code of `wait for` should be contained in `createShare`,  keep the original function definition

File: pkg/sfs/controllerserver.go [hunk 7]
```diff
@@ -40,46 +45,64 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+		VolumeId:      share.ID,
+		ContentSource: req.GetVolumeContentSource(),
+		CapacityBytes: int64(sizeInGiB) * bytesInGiB,
+	}
+	shareIDGetPvcID[share.ID] = req.Name
+	pvcProcessSuccess[req.Name] = volume
+
+	//wait for sfs available
+	share, err = waitForShareStatus(client, share.ID)
```
Comment (chengxiangdong) [inline pkg/sfs/controllerserver.go:92]: do not change the original function definition.

File: pkg/sfs/share.go [hunk 8]
```diff
@@ -57,13 +48,13 @@ func deleteShare(client *golangsdk.ServiceClient, shareID string) error {
 }
 
 // waitForShareStatus wait for share desired status until timeout
-func waitForShareStatus(client *golangsdk.ServiceClient, shareID string, desiredStatus string, timeout int) error {
-	return golangsdk.WaitFor(timeout, func() (bool, error) {
-		share, err := getShare(client, shareID)
+func waitForShareStatus(client *golangsdk.ServiceClient, shareID string) (share *shares.Share, err error) {
```
Comment (chengxiangdong) [inline pkg/sfs/share.go:51]: we do not recommend defining variables in this way
